### PR TITLE
UCT/IB/RC: Combine methods which copy iov and contig buffer to DM to single one.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Re-factored RC/DC am short implementation for device memory.

## Why ?
According to [review comment](https://github.com/openucx/ucx/pull/6200#discussion_r570582887) to reduce code duplication.

## How ?
Replaced implementation of uct_rc_mlx5_iface_common_copy_to_dm by uct_rc_mlx5_iface_copy_iov_to_dm. Because ...copy_to_dm is a special case of ...copy_iov_to_dm.

### CODE REVIEW

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 1]
```diff
@@ -1560,100 +1560,46 @@ done:
...
-}
-
-static void UCS_F_ALWAYS_INLINE uct_rc_mlx5_iface_copy_iov_to_dm(
+static void UCS_F_ALWAYS_INLINE uct_rc_mlx5_iface_common_copy_to_dm(
         uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len, const uct_iov_t *iov,
         size_t iovcnt, void *dm, uct_ib_log_sge_t *log_sge)
 {
     typedef uint64_t misaligned_t UCS_V_ALIGNED(1);
 
+    size_t head = (cache && hdr_len) ? sizeof(cache->bytes) - hdr_len : 0;
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5.inl:1569]: wrap by `()` and use direct comparisonw with `NULL` and `0`:
```
size_t head = ((cache != NULL) && (hdr_len != 0)) ? (sizeof(cache->bytes) - hdr_len) : 0;
```

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 2]
```diff
@@ -1560,100 +1560,46 @@ done:
...
-    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->bytes));
     UCS_STATIC_ASSERT(ucs_static_array_size(log_sge->sg_list) >= 2);
 
-    /* atomically by 8 bytes copy data to DM */
-    /* cache buffer must be aligned, so, source data type is aligned */
-    UCS_WORD_COPY(volatile uint64_t, dm, uint64_t, cache->bytes,
-                  sizeof(cache->bytes));
-    dm = UCS_PTR_BYTE_OFFSET(dm, sizeof(cache->bytes));
+    /* condition is static-evaluated */
+    if (cache && hdr_len) {
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5.inl:1586]:
```
(cache != NULL) && (hdr_len != 0)
```

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 3]
```diff
@@ -1560,100 +1560,46 @@ done:
...
-        log_sge->sg_list[0].length = (uint32_t)hdr_len;
-
-        log_sge->sg_list[1].addr   = (uint64_t)iov->buffer;
-        log_sge->sg_list[1].length = (uint32_t)iov->length;
-
-        log_sge->num_sge = 2;
-    } else {
-        log_sge->num_sge = 0;
+        log_sge->sg_list[i].addr   = (uint64_t)iov->buffer;
+        log_sge->sg_list[i].length = (uint32_t)iov->length;
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5.inl:1600]:
```
i++;
```

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
pls squash

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Pull request merged**